# 1. Постановка задачи

На строительство магистрали периодически доставляются материалы, которые со станции железной  
дороги вначале поступают на $3$ промежуточных склада, а затем непосредственно к $5$ объектам магистрали.  
Два первых склада $А1$, $А2$ имеют ограниченную емкость $50[т]$, и поступившие на них грузы расходуются  
полностью. Cклад $А3$ с неограниченной емкостью допускает резервирование грузов от периода к периоду.  
Транспортные расходы (в условных денежных единицах — $[ДЕ]$) на перевозку одной тонны груза со станции  
на промежуточные склады составляют $7$, $4$, $4[ДЕ]$, а со складов к объектам $В1–В5$ представлены  
в табл. $1$; там же приведены потребности объектов (в тоннах) в течение двух периодов.

<table style="display: inline-block">
    <caption>Таблица $1$: Транспортные расходы</caption>
    <tr>
        <th></th>
        <th>$B1$</th>
        <th>$B2$</th>
        <th>$B3$</th>
        <th>$B4$</th>
        <th>$B5$</th>
    </tr>
    <tr>
        <td>$A1$</td>
        <td>$7$</td>
        <td>$2$</td>
        <td>$13$</td>
        <td>$11$</td>
        <td>$11$</td>
    </tr>
    <tr>
        <td>$A2$</td>
        <td>$11$</td>
        <td>$2$</td>
        <td>$1$</td>
        <td>$9$</td>
        <td>$6$</td>
    </tr>
    <tr>
        <td>$A3$</td>
        <td>$8$</td>
        <td>$11$</td>
        <td>$1$</td>
        <td>$13$</td>
        <td>$13$</td>
    </tr>
    <tr>
        <td>$1й\ период$</td>
        <td>$37$</td>
        <td>$26$</td>
        <td>$31$</td>
        <td>$34$</td>
        <td>$16$</td>
    </tr>
    <tr>
        <td>$2й\ период$</td>
        <td>$57$</td>
        <td>$13$</td>
        <td>$2$</td>
        <td>$19$</td>
        <td>$47$</td>
    </tr>
</table>

Рассматривается работа системы в течение двух периодов при условии доставки на станцию по $150[т]$  
груза в каждый из периодов. Требуется определить план перевозок, обеспечивающий минимум стоимости  
доставки грузов для двух периодов. Провести анализ чувствительности плана к изменению цен перевозок  
из первого промежуточного склада.

# 2. Формализация задачи

* Вектор $\overline{c}$ обозначает стоимость перевозок (в $[ДЕ]$),  
где первые $3$ компоненты - стоимость перевозки в промежуточные склады,  
последующие - стоимости перевозок по строкам таблицы, слева направо;
* Вектор $\overline{x}$ обозначает количество (в $[т]$) товара, которое необходимо распределить.  
Компоненты соответствуют компонентам из вектора $c$ по смыслу "откуда-куда";  
* Матрица $G$ обозначает коэффициенты при компонентах $\overline{x}$ в левой части ограничений на неравенства ($\leq$);
* Вектор $\overline{h}$ обозначает коэффициенты в правой части ограничений на неравенства ($\leq$);
* Матрица $A$ обозначает коэффициенты при компонентах $\overline{x}$ в левой части ограничений на равенства ($=$);
* Вектор $\overline{b}$ обозначает коэффициенты в правой части ограничений на равенства ($=$);

$
\overline{c}_{18x1} =
\begin{pmatrix}
7\\
4\\
4\\
7\\
2\\
13\\
11\\
11\\
11\\
2\\
1\\
9\\
6\\
8\\
11\\
1\\
13\\
13
\end{pmatrix}\quad
\overline{x}_{18x1} =
\begin{pmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
x_5\\
x_6\\
x_7\\
x_8\\
x_9\\
x_{10}\\
x_{11}\\
x_{12}\\
x_{13}\\
x_{14}\\
x_{15}\\
x_{16}\\
x_{17}\\
x_{18}
\end{pmatrix}\quad
\overline{h}_{21x1} =
\begin{pmatrix}
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
50\\
50\\
0
\end{pmatrix}\quad
\overline{b}^{(1)}_{8x1} =
\begin{pmatrix}
150\\
37\\
26\\
31\\
34\\
16\\
0\\
0
\end{pmatrix}\quad
\overline{b}^{(2)}_{8x1} =
\begin{pmatrix}
150\\
57\\
13\\
2\\
19\\
47\\
0\\
0
\end{pmatrix}
\\
G_{21x18} =
\begin{pmatrix}
-1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 \\
1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 1 & 1 & 1 & 1 \\
\end{pmatrix}
\\
A_{8x18} =
\begin{pmatrix}
1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 \\
-1 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\
\end{pmatrix} \\\\
%    [0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],  # _4 + _9 + _14 = 37 / 57
%    [0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],  # _5 + _10 + _15 = 26 / 13
%    [0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],  # _6 + _11 + _16 = 31 / 2
%    [0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0],  # _7 + _12 + _17 = 34 / 19
%    [0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1],  # _8 + _13 + _18 = 16 / 47
%    [-1,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0], # _1 = _4 + _5 + _6 + _7 + _8
%    [0,-1,0,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0], # _2 = _9 + _10 + _11 + _12 + _13
\begin{cases}
    f(\overline{x}^{(j)}) = (\overline{c}^T, \overline{x}^{(j)})-4r^{(j - 1)}
    \rightarrow \min                                                           \\\\
    Gx^{(j)} \leq h                                                            \\\\
    Ax^{(j)}=b^{(j)}
\end{cases}
\qquad j = [1,2],
\ r^{(0)} = 0,\ r^{(1)} = \sum \limits_{n=14}^{18} c^T_nx^{(0)}_n
$

# 3. Описание задачи

Для решения задачи была выбрана библиотека `CVXOPT`.

В данной библиотеке присутствует функция `lp` для решения задач ЛП.  
Ей передаём:
* транспортные расходы ($c$); 
* ограничения на неравенства ($G$ и $h$);
* а также ограничения на равенства ($A$ и $b$).  

Вызываем эту функцию с параметром `solver = 'glpk'`, т.е. будем использовать симплекс-метод.  
Сохраняем значение остатка на промежуточном складе $A3$, а также общую стоимость перевозок.

# 4. Решение задачи

In [1]:
import numpy as np
from cvxopt import matrix, printing, solvers

printing.options['dformat'] = '%7.f'

def solve(c, G, h, A, b, r, p, i):
    s       = solvers.lp(c, G, h, A, b, solver = 'glpk')
    x_A     = s['x'][0:3]
    x_A3    = s['x'][13:18]
    x_A[2] -= r[i]
    p.append(s['primal objective'] - r[i] * c[2])
    r.append(s['x'][2] - sum(x_A3))
    
    print(
        f"Period:\t\t{i + 1}\n\n"
        f"\tStatus:\t\t{s['status']}\n"
        f"\tx:\n\t{s['x'][0:3].T}"
        f"\tx (w.o. remainder):\n\t{x_A.T}"
        f"\t{s['x'][3:8].T}\t{s['x'][8:13].T}\t{x_A3.T}"
        f"\tObjective:\t{p[i]}\n"
        f"\tRemainder(A3):\t{r[i + 1]}\n"
    )

c = matrix([
        7., 4,  4,
        7,  2,  13, 11, 11,
        11, 2,  1,  9,  6,
        8,  11, 1,  13, 13
])
G = matrix([
    [-1.,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1],
    [1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,-1,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1]
]).T
h = matrix([
    0., # -x_1  <= 0
    0,  # -x_2  <= 0
    0,  # -x_3  <= 0
    0,  # -x_4  <= 0
    0,  # -x_5  <= 0
    0,  # -x_6  <= 0
    0,  # -x_7  <= 0
    0,  # -x_8  <= 0
    0,  # -x_9  <= 0
    0,  # -x_10 <= 0
    0,  # -x_11 <= 0
    0,  # -x_12 <= 0
    0,  # -x_13 <= 0
    0,  # -x_14 <= 0
    0,  # -x_15 <= 0
    0,  # -x_16 <= 0
    0,  # -x_17 <= 0
    0,  # -x_18 <= 0
    50, # x_1   <= 50
    50, # x_2   <= 50
    0   # -x_3 + x_14 + x_15 + x_16 + x_17 + x_18 <= 0
])
A = matrix([
    [1.,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], # _1 + _2 + _3 = 150 + r
    [0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],  # _4 + _9 + _14 = 37 / 57
    [0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],  # _5 + _10 + _15 = 26 / 13
    [0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],  # _6 + _11 + _16 = 31 / 2
    [0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0],  # _7 + _12 + _17 = 34 / 19
    [0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1],  # _8 + _13 + _18 = 16 / 47
    [-1,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0], # _1 = _4 + _5 + _6 + _7 + _8
    [0,-1,0,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0], # _2 = _9 + _10 + _11 + _12 + _13
]).T
o = np.array([150.])
s = [[37, 26, 31, 34, 16],
     [57, 13, 2,  19, 47]]
e = [0, 0]
r = [0]
p = []

for i in range(0, 2):
    b = matrix(np.concatenate((o + r[i], s[i], e)))
    solve(c, G, h, A, b, r, p, i)

print(f"Overall price:\t\t{sum(p)}")

Period:		1

	Status:		optimal
	x:
	[     26      50      74]
	x (w.o. remainder):
	[     26      50      74]
	[      0      26       0       0       0]
	[      0       0       0      34      16]
	[     37       0      31       0       0]
	Objective:	1459.0
	Remainder(A3):	6.0

Period:		2

	Status:		optimal
	x:
	[     13      50      93]
	x (w.o. remainder):
	[     13      50      87]
	[      0      13       0       0       0]
	[      0       0       0       3      47]
	[     57       0       2      16       0]
	Objective:	1640.0
	Remainder(A3):	18.0

Overall price:		3099.0


# 5. Интерпретация решения

В $1$ период необходимо перевезти из станции:
* $26[т]$ в $A1$;
* $50[т]$ в $A2$;
* $74[т]$ в $A3$.  

Далее, надо вывезти из:
* $A1$: $26[т]$ в $B2$; 
* $A2$: $34[т]$ в $B4$, $16[т]$ в $B5$;
* $A3$: $37[т]$ в $B1$, $31[т]$ в $B3$.  

Суммарная стоимость этих перевозок будет составлять $1459[ДЕ]$.  
При этом на $A3$ останется $6[т]$ товара.  

Во $2$ период необходимо вывезти из станции:
* $13[т]$ в $A1$;
* $50[т]$ в $A2$;
* $87[т]$ в $A3$
(с учётом $6[т]$, оставшихся с $1$ периода, всего на $A3$ будет $93[т]$).  

Далее, надо вывезти из:
* $A1$: $13[т]$ в $B2$; 
* $A2$: $3[т]$ в $B4$, $47[т]$ в $B5$; 
* $A3$: $57[т]$ в $B1$, $2[т]$ в $B3$, $16[т]$ в $B4$.  

Суммарная стоимость этих перевозок будет составлять $1640[ДЕ]$.  
При этом на $A3$ останется $18[т]$ товара.  

Суммарная стоимость перевозок за два периода будет составлять $3099[ДЕ]$.  
После осуществления перевозок на $A3$ останется $18[т]$ товара.

# 6. Исходные данные для анализа чувствительности и их интерпретация

???

In [2]:
# Анализ чувствительности ???

# 7. Исследование решения для ответа на вопросы, указанные в варианте задания.

???